# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: xpcom/io/nsLocalFileWin.cpp
# Commit: abcbfdf1a1ef
# Full Hash: abcbfdf1a1efad1d852db7889ebda328c4e8b73d
# Author: Toshihito Kikuchi <tkikuchi@mozilla.com>
# Date: 2019-11-15 09:53:19
# Regressor Bug: 1567614
# File Overlap Count: 1
# Description:
#   Bug 1588975 - Replace ShellExecuteExW with mozilla::ShellExecuteByExplorer. r=aklotz,asuth
#   
#   The launcher process turns on the `PreferSystem32Images` mitigation policy for
#   the browser process.  Since the mitigation policy is inherited, a process launched
#   by the browser process also has `PreferSystem32Images`.  If an application which
# ==============================================================================

diff -r 36bef03d2b69 -r abcbfdf1a1ef xpcom/io/nsLocalFileWin.cpp
--- a/xpcom/io/nsLocalFileWin.cpp	Thu Nov 14 18:36:28 2019 +0000
+++ b/xpcom/io/nsLocalFileWin.cpp	Thu Nov 14 19:47:29 2019 +0000
@@ -57,6 +57,7 @@
 #include "mozIDOMWindow.h"
 #include "nsPIDOMWindow.h"
 #include "nsIWidget.h"
+#include "mozilla/ShellHeaderOnlyUtils.h"
 #include "mozilla/WidgetUtils.h"
 
 using namespace mozilla;
@@ -3041,16 +3042,12 @@
   }
 
   // use the app registry name to launch a shell execute....
-  SHELLEXECUTEINFOW seinfo;
-  memset(&seinfo, 0, sizeof(seinfo));
-  seinfo.cbSize = sizeof(SHELLEXECUTEINFOW);
-  seinfo.fMask = SEE_MASK_ASYNCOK;
-  seinfo.hwnd = GetMostRecentNavigatorHWND();
-  seinfo.lpVerb = nullptr;
-  seinfo.lpFile = mResolvedPath.get();
-  seinfo.lpParameters = nullptr;
-  seinfo.lpDirectory = nullptr;
-  seinfo.nShow = SW_SHOWNORMAL;
+  _bstr_t execPath(mResolvedPath.get());
+
+  _variant_t args;
+  _variant_t verb(L"open");
+  _variant_t workingDir;
+  _variant_t showCmd(SW_SHOWNORMAL);
 
   // Use the directory of the file we're launching as the working
   // directory. That way if we have a self extracting EXE it won't
@@ -3058,18 +3055,32 @@
   WCHAR workingDirectory[MAX_PATH + 1] = {L'\0'};
   wcsncpy(workingDirectory, mResolvedPath.get(), MAX_PATH);
   if (PathRemoveFileSpecW(workingDirectory)) {
-    seinfo.lpDirectory = workingDirectory;
+    workingDir = workingDirectory;
   } else {
     NS_WARNING("Could not set working directory for launched file.");
   }
 
-  if (ShellExecuteExW(&seinfo)) {
+  // Ask Explorer to ShellExecute on our behalf, as some applications such as
+  // Skype for Business do not start correctly when inheriting our process's
+  // migitation policies.
+  mozilla::LauncherVoidResult shellExecuteOk = mozilla::ShellExecuteByExplorer(
+      execPath, args, verb, workingDir, showCmd);
+  if (shellExecuteOk.isOk()) {
     return NS_OK;
   }
-  DWORD r = GetLastError();
+  DWORD r = shellExecuteOk.inspectErr().AsWin32Error().value();
   // if the file has no association, we launch windows'
   // "what do you want to do" dialog
   if (r == SE_ERR_NOASSOC) {
+    SHELLEXECUTEINFOW seinfo;
+    memset(&seinfo, 0, sizeof(seinfo));
+    seinfo.cbSize = sizeof(SHELLEXECUTEINFOW);
+    seinfo.fMask = SEE_MASK_ASYNCOK;
+    seinfo.hwnd = GetMostRecentNavigatorHWND();
+    seinfo.lpVerb = nullptr;
+    seinfo.lpDirectory = workingDirectory;
+    seinfo.nShow = SW_SHOWNORMAL;
+
     nsAutoString shellArg;
     shellArg.AssignLiteral(u"shell32.dll,OpenAs_RunDLL ");
     shellArg.Append(mResolvedPath);
